nixops/backends/__init__.py

```diff
@@ -182,16 +183,40 @@ def reboot(self, hard=False):
             reboot_command = "systemctl reboot"
         self.run_command(reboot_command, check=False)
         self.state = self.STARTING
-        self.ssh.reset()
+        if reset:
+            self.ssh.reset()
 
     def reboot_sync(self, hard=False):
         """Reboot this machine and wait until it's up again."""
-        self.reboot(hard=hard)
-        self.log_start("waiting for the machine to finish rebooting...")
-        nixops.util.wait_for_tcp_port(self.get_ssh_name(), self.ssh_port, open=False, callback=lambda: self.log_continue("."))
-        self.log_continue("[down]")
-        nixops.util.wait_for_tcp_port(self.get_ssh_name(), self.ssh_port, callback=lambda: self.log_continue("."))
-        self.log_end("[up]")
+
+        # To check when the machine has finished rebooting in a race-free
+        # manner, we compare the output of `last reboot` before and after
+        # the reboot. Once the output has changed, the reboot is done.
+        def get_last_reboot_output():
+            return self.run_command('last reboot --time-format iso | head -n1', capture_stdout=True).rstrip()
```

Comment:
Both the Hetzner rescue system and the actual NixOS system are using `systemd`, so maybe it's a better idea to use `systemd-analyze` because it [fails whenever bootup is not finished](https://github.com/systemd/systemd/blob/v238/src/analyze/analyze.c#L294). At least that would avoid the current vs. last string comparison.

Reply:
I don't think that can work because when NixOS is waiting for nixops keys to be uploaded (so, right here), you'll get

```
# systemd-analyze
Bootup is not yet finished. Please try again later.
```

Reply:
Oh, right... you're correct and after all we don't care about a truly finished reboot but only want to make sure the machine has rebooted *after all*, so `last reboot` should work.